pkg/webhook/tenant_name/validating.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2020 Clastix Labs.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tenantName
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/clastix/capsule/api/v1alpha1"
+	"github.com/clastix/capsule/pkg/webhook"
+)
+
+// +kubebuilder:webhook:path=/validating-v1-tenant-name,mutating=false,failurePolicy=fail,groups="capsule.clastix.io",resources=tenants,verbs=create,versions=v1alpha1,name=tenant.name.capsule.clastix.io
+
+type Webhook struct{}
+
+func (o Webhook) GetHandler() webhook.Handler {
+	return &handler{}
+}
+
+func (o Webhook) GetName() string {
+	return "TenantName"
+}
+
+func (o Webhook) GetPath() string {
+	return "/validating-v1-tenant-name"
+}
+
+type handler struct{}
+
+func (r *handler) OnCreate(ctx context.Context, req admission.Request, clt client.Client, decoder *admission.Decoder) admission.Response {
+	tnt := &v1alpha1.Tenant{}
+	if err := decoder.Decode(req, tnt); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	fmt.Printf("%s", tnt.GetName())
+
+	matched, _ := regexp.MatchString(`^[a-z0-9]([a-z0-9]*[a-z0-9])?$`, tnt.GetName())
```

Comment:
So we are adding an ability to specify `Kind` of tenant owner. 
Current implementation of owner_refrence webhook use following logic to set tenant for a namespace
https://github.com/clastix/capsule/blob/a99153cbe70432d7e21ad5a2610ba737b1ba27d3/pkg/webhook/owner_reference/patching.go#L84-L93
So basically we find all Tenants for a user\group who created a namespace, and select first one from it (and I think that this logic should be rewritten if future because it's a bit strange for me - a good option will be to set `force-tenant-prefix` by default, so all this magic will gone)

But then we had such thing as `force-tenant-prefix` flag. When this flag is set, we can simply deduct tenant name from namespace name, and that is done in a following code in owner_reference webhook:
https://github.com/clastix/capsule/blob/9d845b367852fb19d1e2756fe7e27fd6338dcdf1/pkg/webhook/owner_reference/patching.go#L89-L102

But here another problem arises. If `-` is used in a tenant name, we won't be able to do it, we need to had some complicated logic with multiple splits and lookups, until we find a suitable tenant. So it order to simplify this process, I've added a webhook which enforces Tenant name to use only alphanumeric characters, thus simplifying overall logic

Reply:
#76 created